pkg/occlient/occlient.go

```diff
@@ -377,43 +391,145 @@ func (c *Client) GetImageStreamsNames(namespace string) ([]string, error) {
+func isTagInImageStream(is imagev1.ImageStream, imageTag string) bool {
+	for _, tag := range is.Status.Tags {
+		// look for matching tag
+		if tag.Tag == imageTag {
+			return true
+		}
+	}
+	return false
+}
+
+// GetImageNS returns the imagestream using image details
+func (c *Client) GetImageStream(imageNS string, imageName string, imageTag string) (*imagev1.ImageStream, error) {
+	var err error
+	var imageStream *imagev1.ImageStream
+	var currentProjectName string
+	currentProjectName = c.GetCurrentProjectName()
+	/*
+		If User has not chosen image NS then,
+			1. Use image from current NS if available
+			2. If not 1, use default openshift NS
+			3. If not 2, return errors from both 1 and 2
+		else
+			Use user chosen namespace
+			If image doesn't exist in user chosen namespace,
+				error out
+			else
+				Proceed
+	*/
+	// User has not passed any particular ImageStream
+	if imageNS == "" {
+
+		// First try finding imagestream from current namespace
+		currentNSImageStream, e := c.imageClient.ImageStreams(currentProjectName).Get(imageName, metav1.GetOptions{})
+		if e != nil {
+			err = errors.Wrapf(e, "no match found for : %s in namespace %s", imageName, currentProjectName)
+		} else {
+			if isTagInImageStream(*currentNSImageStream, imageTag) {
+				return currentNSImageStream, nil
+			}
+		}
+
+		// If not in current namespace, try finding imagestream from openshift namespace
+		openshiftNSImageStream, e := c.imageClient.ImageStreams(OpenShiftNameSpace).Get(imageName, metav1.GetOptions{})
+		if e != nil {
+			// The image is not available in current Namespace.
+			err = errors.Wrapf(e, "%s\n.no match found for : %s in namespace %s", err.Error(), imageName, OpenShiftNameSpace)
+		} else {
+			if isTagInImageStream(*openshiftNSImageStream, imageTag) {
+				return openshiftNSImageStream, nil
+			}
+		}
+		if e != nil && err != nil {
+			// Imagestream not found in openshift and current namespaces
+			return nil, err
+		}
+
+		// Required tag not in openshift and current namespaces
+		return nil, errors.Errorf("image stream %s with tag %s not found in openshift and %s namespaces", imageName, imageTag, currentProjectName)
+
+	} else {
+
+		// Fetch imagestream from requested namespace
+		imageStream, err = c.imageClient.ImageStreams(imageNS).Get(imageName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(
+				err, "no match found for %s in namespace %s", imageName, imageNS,
+			)
+		}
+		if !isTagInImageStream(*imageStream, imageTag) {
+			return nil, errors.Errorf("image stream %s with tag %s not found in %s namespaces", imageName, imageTag, currentProjectName)
+		}
+	}
+
+	return imageStream, nil
+}
+
+// GetExposedPorts retruns image namespace and list of ContainerPorts that are exposed by given image
+func (c *Client) GetExposedPorts(imageNS string, imageName string, imageTag string) (string, []corev1.ContainerPort, error) {
```

Comment:
The imagestream is not expected to always be passed by the user.
This Pr has the capability to resolve imagestream based on:
0. User chosen namespace(if explicitly passed)
1. Priority for the imagestream in current namespace.
2. Availability of required tag in the currently considered namespace
For cases 1 and 2, when user has not passed imagestream namespace, the resolved namespace will need to be set into build config for imagestream mapping although, it can be also resolved by exclusive separate call to  GetImageStream, in order to avoid multiple calls to this(GetImageStream) function, I made GetExposedPorts also return namespace

Reply:
It is weird and not really transparent when the function has the same argument that is also returned.
Especially when the function's responsibility is to just return ports. I think that function like this should expect that `imageNS` is always explicitly passed.

Reply:
Ack..
But adds a call extra(2 calls everywhere(where GetExposedPorts was used) instead of one previously) :)